src/main/kotlin/com/openlattice/mechanic/integrity/OrphanedLinkingPropertyData.kt

```diff
@@ -69,47 +70,76 @@ class OrphanedLinkingPropertyData(private val toolbox: Toolbox) : Check {
     private fun deleteOrphanedLinkingProperties() {
         logger.info("Starting task to delete orphaned linking property types values.")
         val sw = Stopwatch.createStarted()
-
-        do {
-            sw.reset().start()
-            var count = 0L
-
-            BasePostgresIterable(
-                    StatementHolderSupplier(toolbox.hds, selectDeletableSql)
-            ) { rs ->
-                LinkedEntityRow(
-                        ResultSetAdapters.id(rs), ResultSetAdapters.originId(rs), ResultSetAdapters.propertyTypeId(rs)
-                )
-            }
-                    .groupBy { it.linkingId }
-                    .forEach { (linkingId, deletablesOfLinkingId) ->
-                        deletablesOfLinkingId
-                                .groupBy { it.originId }
-                                .forEach { (originId, deletablesOfOriginId) ->
-                                    count += toolbox.hds.connection.use { conn ->
-                                        val propertyTypeIdsArr = PostgresArrays.createUuidArray(
-                                                conn,
-                                                deletablesOfOriginId.map { it.propertyTypeId }
-                                        )
-                                        conn.prepareStatement(deleteSql).use {
-                                            it.setObject(1, linkingId)
-                                            it.setObject(2, originId)
-                                            it.setArray(3, propertyTypeIdsArr)
-                                            it.executeUpdate()
-                                        }
-                                    }
-                                }
+        val swTotal = Stopwatch.createStarted()
+        var countTotal = 0L
+
+        selectAllLinkingIds()
```

Comment:
I think it might be worth also selecting `array_agg( DISTINCT ${ENTITY_SET_ID.name} )` and `array_agg( DISTINCT ${PARTITION.name} )` in this initial query, and then using those as extra filters within `selectDeletableSql` and `deleteSql` -- might not help a ton depending how many there are in each batch but then at least postgres will be able to use its data table indices to speed this up a bit

Reply:
***actually, assuming we're able to load these ids all into memory anyway, it's probably even better to do a preprocessing step after loading them all, mapping `partition -> linkingId -> (originIds, entitySetIds)` and then do a separate query for each partition (potentially batching within each partition depending on how many entries there are)

Reply:
That sounds like a good idea, but I'm concerned about matching partitions for linking entries and their original entries. Now, we partition by originId for linking entities, but this wasn't the case always and I'm not sure, if we re-partitioned those rows in data table (or ids table). Because if not and we try to match the same partitions in the inner and outer query (second query), it will return those linking entries, where the original entries are there, but with different partitions and will delete those..

If we are sure, that we did the repartitioning, I think, it's enough to group by partition and do a batch call by each partition.

Reply:
I'm not sure if we did the repartitioning, but if we do an `array_agg( DISTINCT ${PARTITION.name} )` in the initial query won't that bring back every partition that's actually in use for every linking_id in the data table? And since we'd have to assume there are multiple partitions per linking_id anyway (because of multiple normal entities that comprise a linking entity) it shouldn't make a difference in the approach

Reply:
Let's say we have these rows in data table (which shouldn't be deleted, since both linking entries and original entries are present):
```
ID | ORIGIN_ID | ES_ID | PART
l1     o1        e1       p1  // this entry was before partitioning by origin_id
l1     o2        e1       p2
o1     -         e1       p2
o2     -         e1       p2 
```

When we would do the second query with values `p1, l1, o1, e1`, the inner query would return an empty result for
```
(SELECT id,
            property_type_id
     FROM DATA
     WHERE origin_id = '7fffffff-ffff-ffff-7fff-ffffffffffff'
       AND id = ANY(o1)
       AND partition = ANY(p1)
       AND entity_set_id = ANY(e1)
       AND VERSION > 0)
```
And we do a `NOT IN` with the outer entries (which is basically the first row in the table) and that would result with `l1, o1` (and property type ids)

Reply:
If we don't filter on partition in the inner query, it should work fine, but the original problem was, that the inner query was too big

Reply:
Yeah, or just filter on all partitions for all linking ids in the batch. Would require some preprocessing but that overhead is probably way less than extra avoidable scans of the data table.

Reply:
Cool, will change i!